File: sdk/go/login/audience.go
package login

import (
	"encoding/hex"
	"errors"
	"fmt"
	"strconv"
	"strings"
)

// DeriveAudience turns the URL a person is told to run — `<audience>/login/<id>` — into the
// audience the login proof binds and the request id's bytes (docs/login.md §2.1).
//
// It is part of the scheme, not of any CLI, because three URL parsers normalise three ways
// and the audience is the first field of the binding: a WHATWG URL drops a default port,
// net/url keeps it and decodes the path, a hand-rolled parser keeps userinfo. One URL must
// yield one audience in every lane, so this function accepts exactly the grammar in §2.1
// and REFUSES everything else rather than normalising it:
//
//	invocation = scheme "://" host [ ":" port ] *( "/" segment ) "/login/" id
//
// scheme is http/https/ws/wss (case-insensitive; ws → http, wss → https); host is an ASCII
// reg-name or a bracketed IPv6 literal, lowercased; a port is 1..=65535 without a leading
// zero and is omitted from the audience when it is the folded scheme's default (80, 443);
// segments are kept as written (case preserved, percent-escapes never decoded); the id is
// lowercase hex of even length and is returned decoded. Any non-ASCII, whitespace or
// control byte, userinfo, query, fragment, empty segment, `.`/`..` segment, malformed
// escape, or a penultimate segment other than `login` is an error.
func DeriveAudience(rawURL string) (audience string, id []byte, err error) {
	for i := 0; i < len(rawURL); i++ {
		if c := rawURL[i]; c <= 0x20 || c >= 0x7f {
			return "", nil, fmt.Errorf("login: URL byte %d is 0x%02x — only printable ASCII is accepted", i, c)
		}
	}
	sep := strings.Index(rawURL, "://")
	if sep < 0 {
		return "", nil, errors.New("login: URL has no scheme")
	}
	scheme, defaultPort, err := foldScheme(rawURL[:sep])
	if err != nil {
		return "", nil, err
	}
	rest := rawURL[sep+3:]
	slash := strings.IndexByte(rest, '/')
	if slash < 0 {
		return "", nil, errors.New("login: URL has no path — expected <audience>/login/<id>")
	}
	host, port, err := parseAuthority(rest[:slash])
	if err != nil {
		return "", nil, err
	}
	path := rest[slash:]
	if strings.ContainsAny(path, "?#") {
		return "", nil, errors.New("login: URL carries a query or fragment — expected <audience>/login/<id>")
	}
	segments := strings.Split(path[1:], "/")
	for i, seg := range segments {
		if err := checkSegment(seg); err != nil {
			return "", nil, fmt.Errorf("login: path segment %d: %w", i, err)
		}
	}
	if len(segments) < 2 {
		return "", nil, errors.New("login: URL is not <audience>/login/<id>")
	}
	if segments[len(segments)-2] != "login" {
		return "", nil, fmt.Errorf("login: expected <audience>/login/<id>, got %q as the penultimate segment", segments[len(segments)-2])
	}
	id, err = decodeID(segments[len(segments)-1])
	if err != nil {
		return "", nil, err
	}
	var b strings.Builder
	b.WriteString(scheme)
	b.WriteString("://")
	b.WriteString(host)
	if port != "" && port != defaultPort {
		b.WriteByte(':')
		b.WriteString(port)
	}
	for _, seg := range segments[:len(segments)-2] {
		b.WriteByte('/')
		b.WriteString(seg)
	}
	return b.String(), id, nil
}

// foldScheme lowercases and folds the scheme; returns it with the default port it implies.
func foldScheme(s string) (scheme, defaultPort string, err error) {
	switch strings.ToLower(s) {
	case "http", "ws":
		return "http", "80", nil
	case "https", "wss":
		return "https", "443", nil
	}
	return "", "", fmt.Errorf("login: unsupported URL scheme %q — want http(s) or ws(s)", s)
}

// parseAuthority lowercases the host and validates the port; userinfo is refused.
func parseAuthority(auth string) (host, port string, err error) {
	if strings.IndexByte(auth, '@') >= 0 {
		return "", "", errors.New("login: URL carries userinfo — refused")
	}
	if auth == "" {
		return "", "", errors.New("login: URL has no host")
	}
	if auth[0] == '[' {
		end := strings.IndexByte(auth, ']')
		if end < 0 {
			return "", "", errors.New("login: unterminated IPv6 literal")
		}
		host = strings.ToLower(auth[:end+1])
		lit := host[1:end]
		if len(lit) < 2 || !strings.Contains(lit, ":") {
			return "", "", errors.New("login: malformed IPv6 literal")
		}
		for i := 0; i < len(lit); i++ {
			if c := lit[i]; !(isHexLower(c) || c == ':' || c == '.') {
				return "", "", fmt.Errorf("login: IPv6 literal has byte 0x%02x — refused", c)
			}
		}
		tail := auth[end+1:]
		if tail == "" {
			return host, "", nil
		}
		if tail[0] != ':' {
			return "", "", errors.New("login: bytes after the IPv6 literal — refused")
		}
		port, err = checkPort(tail[1:])
		return host, port, err
	}
	name := auth
	if colon := strings.IndexByte(auth, ':'); colon >= 0 {
		name = auth[:colon]
		if port, err = checkPort(auth[colon+1:]); err != nil {
			return "", "", err
		}
	}
	host = strings.ToLower(name)
	for _, label := range strings.Split(host, ".") {
		if label == "" {
			return "", "", errors.New("login: host has an empty label — refused")
		}
		for i := 0; i < len(label); i++ {
			if c := label[i]; !((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '-') {
				return "", "", fmt.Errorf("login: host has byte 0x%02x — refused (IDNs must be given as punycode)", c)
			}
		}
	}
	return host, port, nil
}

// checkPort accepts 1..=65535 written without a leading zero.
func checkPort(p string) (string, error) {
	if p == "" || (len(p) > 1 && p[0] == '0') {
		return "", fmt.Errorf("login: port %q — refused", p)
	}
	for i := 0; i < len(p); i++ {
		if p[i] < '0' || p[i] > '9' {
			return "", fmt.Errorf("login: port %q is not a number", p)
		}
	}
	n, err := strconv.Atoi(p)
	if err != nil || n < 1 || n > 65535 {
		return "", fmt.Errorf("login: port %q is out of range", p)
	}
	return p, nil
}

// checkSegment is RFC 3986 pchar with escapes kept as written; `.` and `..` are refused.
func checkSegment(seg string) error {
	if seg == "" {
		return errors.New("empty segment (a trailing slash or `//`) — refused")
	}
	if seg == "." || seg == ".." {
		return errors.New("dot segment — refused")
	}
	for i := 0; i < len(seg); i++ {
		c := seg[i]
		switch {
		case (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9'):
		case strings.IndexByte("-._~!$&'()*+,;=:@", c) >= 0:
		case c == '%':
			if i+2 >= len(seg) || !isHex(seg[i+1]) || !isHex(seg[i+2]) {
				return errors.New("malformed percent-escape — refused")
			}
			i += 2
		default:
			return fmt.Errorf("byte 0x%02x is not allowed in a path segment — refused", c)
		}
	}
	return nil
}

// decodeID accepts lowercase hex of even length ≥ 2 and returns the bytes.
func decodeID(s string) ([]byte, error) {
	if len(s) < 2 || len(s)%2 != 0 {
		return nil, fmt.Errorf("login: id %q is not lowercase hex of even length", s)
	}
	for i := 0; i < len(s); i++ {
		if !isHexLower(s[i]) {
			return nil, fmt.Errorf("login: id %q is not lowercase hex of even length", s)
		}
	}
	return hex.DecodeString(s)
}

func isHexLower(c byte) bool { return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') }
func isHex(c byte) bool      { return isHexLower(c) || (c >= 'A' && c <= 'F') }

File: sdk/rs/src/login/audience.rs
//! `derive_audience` — the invocation URL → (audience, id bytes), `docs/login.md` §2.1.
//!
//! Part of the scheme, not of any CLI: three URL parsers normalise three ways and the audience
//! is the first field of the binding, so one URL must yield one audience in every lane. This
//! function accepts exactly the grammar of §2.1 and **refuses everything else rather than
//! normalising it** — no URL library, on purpose (a library is a normaliser, and the `url`
//! crate alone would break the sdk's one-dependency rule).
//!
//! ```text
//! invocation = scheme "://" host [ ":" port ] *( "/" segment ) "/login/" id
//! ```

/// The audience the login proof binds and the request id's bytes, from the URL the person
/// was told to run. Scheme is `http`/`https`/`ws`/`wss` (case-insensitive; `ws` → `http`,
/// `wss` → `https`); the host is an ASCII reg-name or a bracketed IPv6 literal, lowercased; a
/// port is `1..=65535` without a leading zero and is omitted from the audience when it is the
/// folded scheme's default (80, 443); segments are kept as written (case preserved,
/// percent-escapes never decoded); the id is lowercase hex of even length and is returned
/// decoded. Any non-ASCII, whitespace or control byte, userinfo, query, fragment, empty
/// segment, `.`/`..` segment, malformed escape, or a penultimate segment other than `login` is
/// an error.
pub fn derive_audience(url: &str) -> Result<(String, Vec<u8>), String> {
    if let Some((i, c)) = url
        .bytes()
        .enumerate()
        .find(|(_, c)| *c <= 0x20 || *c >= 0x7f)
    {
        return Err(format!(
            "login: URL byte {i} is 0x{c:02x} — only printable ASCII is accepted"
        ));
    }
    let (raw_scheme, rest) = url
        .split_once("://")
        .ok_or_else(|| "login: URL has no scheme".to_string())?;
    let (scheme, default_port) = fold_scheme(raw_scheme)?;
    let (authority, path) = match rest.find('/') {
        Some(i) => (&rest[..i], &rest[i..]),
        None => return Err("login: URL has no path — expected <audience>/login/<id>".to_string()),
    };
    let (host, port) = parse_authority(authority)?;
    if path.contains('?') || path.contains('#') {
        return Err(
            "login: URL carries a query or fragment — expected <audience>/login/<id>".to_string(),
        );
    }
    let segments: Vec<&str> = path[1..].split('/').collect();
    for (i, seg) in segments.iter().enumerate() {
        check_segment(seg).map_err(|e| format!("login: path segment {i}: {e}"))?;
    }
    if segments.len() < 2 {
        return Err("login: URL is not <audience>/login/<id>".to_string());
    }
    let penultimate = segments[segments.len() - 2];
    if penultimate != "login" {
        return Err(format!(
            "login: expected <audience>/login/<id>, got {penultimate:?} as the penultimate segment"
        ));
    }
    let id = decode_id(segments[segments.len() - 1])?;
    let mut audience = format!("{scheme}://{host}");
    if !port.is_empty() && port != default_port {
        audience.push(':');
        audience.push_str(port);
    }
    for seg in &segments[..segments.len() - 2] {
        audience.push('/');
        audience.push_str(seg);
    }
    Ok((audience, id))
}

fn fold_scheme(s: &str) -> Result<(&'static str, &'static str), String> {
    match s.to_ascii_lowercase().as_str() {
        "http" | "ws" => Ok(("http", "80")),
        "https" | "wss" => Ok(("https", "443")),
        _ => Err(format!(
            "login: unsupported URL scheme {s:?} — want http(s) or ws(s)"
        )),
    }
}

fn parse_authority(auth: &str) -> Result<(String, &str), String> {
    if auth.contains('@') {
        return Err("login: URL carries userinfo — refused".to_string());
    }
    if auth.is_empty() {
        return Err("login: URL has no host".to_string());
    }
    if let Some(stripped) = auth.strip_prefix('[') {
        let end = stripped
            .find(']')
            .ok_or_else(|| "login: unterminated IPv6 literal".to_string())?;
        let lit = stripped[..end].to_ascii_lowercase();
        if lit.len() < 2
            || !lit.contains(':')
            || !lit
                .bytes()
                .all(|c| c.is_ascii_hexdigit() && !c.is_ascii_uppercase() || c == b':' || c == b'.')
        {
            return Err("login: malformed IPv6 literal".to_string());
        }
        let tail = &stripped[end + 1..];
        let port = if tail.is_empty() {
            ""
        } else {
            let p = tail
                .strip_prefix(':')
                .ok_or_else(|| "login: bytes after the IPv6 literal — refused".to_string())?;
            check_port(p)?
        };
        return Ok((format!("[{lit}]"), port));
    }
    let (name, port) = match auth.split_once(':') {
        Some((n, p)) => (n, check_port(p)?),
        None => (auth, ""),
    };
    let host = name.to_ascii_lowercase();
    for label in host.split('.') {
        if label.is_empty() {
            return Err("login: host has an empty label — refused".to_string());
        }
        if let Some(c) = label
            .bytes()
            .find(|c| !(c.is_ascii_lowercase() || c.is_ascii_digit() || *c == b'-'))
        {
            return Err(format!(
                "login: host has byte 0x{c:02x} — refused (IDNs must be given as punycode)"
            ));
        }
    }
    Ok((host, port))
}

fn check_port(p: &str) -> Result<&str, String> {
    if p.is_empty() || (p.len() > 1 && p.starts_with('0')) || !p.bytes().all(|c| c.is_ascii_digit())
    {
        return Err(format!("login: port {p:?} — refused"));
    }
    match p.parse::<u32>() {
        Ok(n) if (1..=65535).contains(&n) => Ok(p),
        _ => Err(format!("login: port {p:?} is out of range")),
    }
}

fn check_segment(seg: &str) -> Result<(), String> {
    if seg.is_empty() {
        return Err("empty segment (a trailing slash or `//`) — refused".to_string());
    }
    if seg == "." || seg == ".." {
        return Err("dot segment — refused".to_string());
    }
    let b = seg.as_bytes();
    let mut i = 0;
    while i < b.len() {
        let c = b[i];
        if c.is_ascii_alphanumeric() || b"-._~!$&'()*+,;=:@".contains(&c) {
            i += 1;
        } else if c == b'%' {
            if i + 2 >= b.len() || !b[i + 1].is_ascii_hexdigit() || !b[i + 2].is_ascii_hexdigit() {
                return Err("malformed percent-escape — refused".to_string());
            }
            i += 3;
        } else {
            return Err(format!(
                "byte 0x{c:02x} is not allowed in a path segment — refused"
            ));
        }
    }
    Ok(())
}

fn decode_id(s: &str) -> Result<Vec<u8>, String> {
    let b = s.as_bytes();
    if b.len() < 2
        || !b.len().is_multiple_of(2)
        || !b
            .iter()
            .all(|c| c.is_ascii_digit() || (b'a'..=b'f').contains(c))
    {
        return Err(format!(
            "login: id {s:?} is not lowercase hex of even length"
        ));
    }
    Ok(b.chunks(2)
        .map(|pair| {
            let hi = (pair[0] as char).to_digit(16).unwrap() as u8;
            let lo = (pair[1] as char).to_digit(16).unwrap() as u8;
            (hi << 4) | lo
        })
        .collect())
}

#[cfg(test)]
mod tests {
    use super::derive_audience;

    const ID: &str = "8f3c1d2e4b5a69780f1e2d3c4b5a6978";

    #[test]
    fn accepts() {
        let cases = [
            (
                format!("https://dawn.example/api/login/{ID}"),
                "https://dawn.example/api",
            ),
            (
                format!("https://dawn.example:443/api/login/{ID}"),
                "https://dawn.example/api",
            ),
            (
                format!("http://localhost:8080/login/{ID}"),
                "http://localhost:8080",
            ),
            (
                format!("HTTPS://Dawn.Example/API/login/{ID}"),
                "https://dawn.example/API",
            ),
            (
                format!("wss://dawn.example:80/login/{ID}"),
                "https://dawn.example:80",
            ),
            (
                format!("https://[2001:DB8::1]:443/login/{ID}"),
                "https://[2001:db8::1]",
            ),
            (
                format!("https://dawn.example/a/b%2Fc/login/{ID}"),
                "https://dawn.example/a/b%2Fc",
            ),
        ];
        for (url, want) in cases {
            let (aud, id) = derive_audience(&url).unwrap_or_else(|e| panic!("{url}: {e}"));
            assert_eq!(aud, want, "{url}");
            assert_eq!(id.len(), 16);
        }
    }

    #[test]
    fn refuses() {
        for url in [
            format!("dawn.example/login/{ID}"),
            format!("ftp://dawn.example/login/{ID}"),
            format!("https://dawn.example/api/login/{ID}/"),
            format!("https://user@dawn.example/login/{ID}"),
            format!("https://dawn.example/login/{ID}?x=1"),
            format!("https://dawn.example:0443/login/{ID}"),
            "https://dawn.example/api/login/8F3C".to_string(),
            "https://dawn.example/api/login/8f3".to_string(),
            format!("https://dawn.example/a b/login/{ID}"),
            format!("https://dawn.example/a%2/login/{ID}"),
            format!("https://[fe80::1%25eth0]/login/{ID}"),
            String::new(),
        ] {
            assert!(derive_audience(&url).is_err(), "{url}: not refused");
        }
    }
}
